### ISSUE TITLE
lstat error when requiring virtual modules

### ISSUE THREAD

**Issue by craigbeck** [webpack/enhanced-resolve#411]
Latest v5.16.0 breaks with `lstat` error when using virtual modules:

```
ERROR in ../../../../derby-webpack-virtual-fs/app-views/ sync ^\.\/.*__views$
ENOENT: no such file or directory, lstat '/derby-webpack-virtual-fs'
caused by plugins in ContextModule.resolveDependencies
Error: ENOENT: no such file or directory, lstat '/derby-webpack-virtual-fs'
 @ ./node_modules/@derbyjs/derby-webpack/lib/browser.js 8:11-75
```

We use a plugin that writes virtual modules (using `webpack-virtual-modules`) [code](https://github.com/derbyjs/derby-webpack/blob/v3.3.2/lib/DerbyViewPlugin.js#L42-L46), but fails to `require` these modules [here](https://github.com/derbyjs/derby-webpack/blob/v3.3.2/lib/browser.js#L8)

This works if we pin to prior v5.15.1 version.

**Comment by alexander-akait** [comment]
We don't change `lstat` logic, please create reproducible test repo

**Comment by dobladov** [comment]
I also have the same issue with `v5.16.0` and virtual modules, the only change I could see is that in the previous version `lstat` was called as `lstatSync.call(fs, arg);` without options, as opposed to the newer version where `lstatSync.call(fs, arg, options)` is called if a callback is present.

[`v5.15.1...v5.16.1`#diff-0f5c53a719](https://github.com/webpack/enhanced-resolve/compare/v5.15.1...v5.16.1#diff-0f5c53a719f1116101cd5cb9491543d09f542eb8d96251ce019eec779cc070a6R30-R32)

**Comment by alexander-akait** [comment]
@dobladov Can you create reproducible test repo?

**Comment by dobladov** [comment]
I will try to reproduce the issue, it happens in conjunction with [webpack-virtual-modules](https://www.npmjs.com/package/webpack-virtual-modules) when trying to use `writeModule`:

```js
const VirtualModulesPlugin = require('webpack-virtual-modules');

const virtualModules = new VirtualModulesPlugin();

virtualModules.writeModule(
  'node_modules/module-foo.js',
  'module.exports = { foo: "foo" };'
);
```

**Comment by alexander-akait** [comment]
@dobladov Sorry I still can't reproduce, my configuration:
```
const VirtualModulesPlugin = require('webpack-virtual-modules');

module.exports = {
	entry: "./src/entry-1.js",
	mode: "development",
	target: "node",
	devtool: false,
	experiments: {
		css: true
	},
	plugins: [
		new VirtualModulesPlugin({
			"node_modules/module-foo.js": 'module.exports = { foo: "foo" };'
		})
	]
};

```

**Comment by dobladov** [comment]
@alexander-akait I was also struggling to reproduce it, so I'm not sure what might fail, thanks for attempting @alexander-akait

**Comment by alexander-akait** [comment]
Reproduced

**Comment by alexander-akait** [comment]
Example:
```js
const VirtualModulesPlugin = require('webpack-virtual-modules');

module.exports = {
	entry: "./src/entry-1.js",
	mode: "development",
	target: "node",
	devtool: false,
	experiments: {
		css: true
	},
	plugins: [
		new VirtualModulesPlugin({
			"node_modules/foo/module-foo.js": 'module.exports = { foo: "foo" };'
		})
	]
};
```

```
function test() {
	const myName = 1;

	import(`foo/${myName}`);
}
```

**Comment by alexander-akait** [comment]
The problem is here - https://github.com/sysgears/webpack-virtual-modules/blob/master/src/index.ts#L240, `webpack-virtual-modules` tries to emulate `lstats` (and stat logic), but because now we cache `fs.realpath` for perf reasons webpack tries to use it, but `webpack-virtual-modules` doesn't emulate `realpath` logic, so you got this error.

There are two solution:
1. I can fallback on old behaviour in webpack we we failed with `realpath`, it is like a hack solution then normal
2. Fix it in `webpack-virtual-modules` and emulate realpath logic too (if `realpath` is avaliable on backend filesystem)

**Comment by alexander-akait** [comment]
@dobladov there is a fix - https://github.com/sysgears/webpack-virtual-modules/pull/175, I can't merge, so we need to wait

**Comment by alexander-akait** [comment]
No problems with `enhanced-resolve` here, so I want to close it here, anyway feel free to feedback

**Comment by dobladov** [comment]
@alexander-akait thanks for your hard work in this fix, that was a very fast and clear response.

**Comment by larixer** [comment]
@alexander-akait @dobladov The fix has been merged and published in `webpack-virtual-modules@0.6.2`
